orderbook/src/bad_token.rs

```diff
@@ -0,0 +1,465 @@
+                        EthcontractError::Node => Err(err.into()),
+                        EthcontractError::Contract => Ok(None),
+                    }
+                }
+            };
+            match biggest_balance {
+                Some((_, current_biggest)) if current_biggest > balance => (),
+                _ => biggest_balance = Some((address, balance)),
+            }
+        }
+        Ok(biggest_balance)
+    }
+
+    fn create_trace_request(&self, token: H160, amount: U256, take_from: H160) -> Vec<CallRequest> {
+        let instance = ERC20::at(&self.web3, token);
+
+        // We transfer one unit (base on decimals) of the token from the amm pool into the
+        // settlement contract and then to an arbitrary address.
+        // Note that gas use can depend on the recipient because for the standard implementation
+        // sending to an address that does not have any balance yet (implicitly 0) causes an
+        // allocation.
+
+        let mut requests = Vec::new();
+
+        // 0
+        let tx = instance.balance_of(self.settlement_contract).m.tx;
+        requests.push(call_request(None, token, tx));
+        // 1
+        let tx = instance.transfer(self.settlement_contract, amount).tx;
+        requests.push(call_request(Some(take_from), token, tx));
+        // 2
+        let tx = instance.balance_of(self.settlement_contract).m.tx;
+        requests.push(call_request(None, token, tx));
+        // 3
+        // For the out transfer we use an arbitrary address without balance to detect tokens that
+        // usually apply fees but not if the the sender or receiver is specifically exempt like
+        // their own uniswap pools.
+        let recipient = H160::from_low_u64_le(1);
+        let tx = instance.transfer(recipient, amount).tx;
+        requests.push(call_request(Some(self.settlement_contract), token, tx));
+        // 4
+        let tx = instance.balance_of(self.settlement_contract).m.tx;
+        requests.push(call_request(None, token, tx));
+        // 5
+        let tx = instance.balance_of(recipient).m.tx;
+        requests.push(call_request(None, token, tx));
+
+        requests
+    }
+
+    fn handle_response(&self, traces: &[BlockTrace], amount: U256) -> Result<TokenInfo> {
+        ensure!(traces.len() == 6, "unexpected number of traces");
+
+        let gas_in = match ensure_transaction_ok_and_get_gas(&traces[1])? {
+            Ok(gas) => gas,
+            Err(reason) => return Ok(TokenInfo::Bad { reason }),
+        };
+        let gas_out = match ensure_transaction_ok_and_get_gas(&traces[3])? {
+            Ok(gas) => gas,
+            Err(reason) => return Ok(TokenInfo::Bad { reason }),
+        };
+
+        let balance_before_in = match decode_u256(&traces[0]) {
+            Ok(balance) => balance,
+            Err(_) => {
+                return Ok(TokenInfo::Bad {
+                    reason: "can't decode balance".to_string(),
```

Comment:
nit: maybe qualify which balance can't be decoded:
- `can't decode initial settlement balance`
- `can't decode final settlement balance`
- `can't decode initial recipient balance`
- `can't decode final recipient balance`